### ISSUE TITLE
TypeError in your package ? 

### ISSUE THREAD

**Issue by Thebigbignooby** [dimadeveatii/mongo-dot-notation#6]
```
{ id: '58751ab0b01f73441d416c3b',
  physicalDescription: 
   { title: 'description title',
     location: { address: 'aaaaaaaaaa', city: 'aaaaaaaaaa', country: 'France' },
     owners: [ [Object] ],
     kind: 'house',
     surface: { total: 69, effective: 50 } } }
```
I tried using this, however . . . 
```
TypeError: Cannot read property 'name' of undefined
    at isPrimitive (/home/rory/dev/weezhome/api/node_modules/mongo-dot-notation/index.js:53:19)
    at flatten (/home/rory/dev/weezhome/api/node_modules/mongo-dot-notation/index.js:19:7)
    at /home/rory/dev/weezhome/api/node_modules/mongo-dot-notation/index.js:43:5
    at Array.forEach (native)
    at flatten (/home/rory/dev/weezhome/api/node_modules/mongo-dot-notation/index.js:41:8)
    at module.exports.flatten (/home/rory/dev/weezhome/api/node_modules/mongo-dot-notation/index.js:14:3)
    at _callee5$ (/home/rory/dev/weezhome/api/src/schema/property.js:174:24)
    at tryCatch (/home/rory/dev/weezhome/api/node_modules/regenerator-runtime/runtime.js:64:40)
    at GeneratorFunctionPrototype.invoke [as _invoke] (/home/rory/dev/weezhome/api/node_modules/regenerator-runtime/runtime.js:355:22)
    at GeneratorFunctionPrototype.prototype.(anonymous function) [as next] (/home/rory/dev/weezhome/api/node_modules/regenerator-runtime/runtime.js:116:21)
```

**Comment by dimadeveatii** [comment]
Hi @Thebigbignooby. Could you paste please the code snippet to reproduce this issue?

**Comment by Thebigbignooby** [comment]
I've moved on from the moment I had this error. But the code was straightforward, something like this

```javascript
import { flatten } from 'mongo-dot-notation'

// graphql mutation resolver where args is the object I pasted at the top of this issue
async function updateProperty (root, args, context, info) {
  if (!context.viewer) return null
  const { id, ...nestedSet } = args
  const set = flatten(nestedSet)
  const fields = getProjection(info)
  const userQuery = (isAdmin(context.viewer) ? {} : { userId: new ObjectID(context.viewer.userId) })
  const property = await context.Properties.findOneAndUpdate(
    { _id: new ObjectID(id), ...userQuery }, { $set: set }, { returnOriginal: false, fields }
  )
  return property ? { ...property.value, id: property.value._id } : null
} 
```

it wasn't exactly this code but essentially that's how I used your package

I ended up removing mongo-dot-notation and instead used [dot-object](https://github.com/rhalff/dot-object)

although I now have an issue where my arrays become objects in mongo as a consequence. does your library prevent this from happening?

**Comment by dimadeveatii** [comment]
I believe this happens because the object (or a nested property) has a null constructor. I think that is normally not the case with usual JS types, unless you have some library that enables inheritance and uses null for a base class constructor (instead of object).

I've just applied a fix and published *1.0.7* version on npm which handles this scenario.

As with your question about arrays, they are handled and not touched. If you pass an array, it will be set also as an array. This is covered in unit tests.

Please let me know whether latest version fixes this issue.

**Comment by Thebigbignooby** [comment]
It should be a normal argument object one receives from a post request in node.js

although in my graphql schema I did in fact have an Owner interface  which was implemented by 2 different types, but that doesn't affect the arguments object.

concerning my question about arrays, I was talking about arrays of objects such as 

```
......,
"rooms": [
      {
        "kind": "toilets",
        "totalSurface": 1,
        "effectiveSurface": 1,
        "chimney": 0,
        "state": "needs_work"
      },
      {
        "kind": "custom",
        "customKind": "office",
        "totalSurface": 20,
        "effectiveSurface": 20,
        "chimney": 0,
        "state": "average"
      }
    ],
 .....
``` 

dot-object transforms it into 
```
  'physicalDescription.rooms.0.kind': 'toilets',
  'physicalDescription.rooms.0.totalSurface': 1,
  'physicalDescription.rooms.0.effectiveSurface': 1,
  'physicalDescription.rooms.0.state': 'needs_work',
  'physicalDescription.rooms.0.chimney': 0,
  'physicalDescription.rooms.1.kind': 'custom',
  'physicalDescription.rooms.1.customKind': 'office',
  'physicalDescription.rooms.1.totalSurface': 20,
  'physicalDescription.rooms.1.effectiveSurface': 20,
  'physicalDescription.rooms.1.state': 'average',
  'physicalDescription.rooms.1.chimney': 0 }
```

in mongoDB rooms would be create as an Object, I solved it by creating empty arrays by default in my graphQL schema

**Comment by dimadeveatii** [comment]
With [mongo-dot-notation](https://www.npmjs.com/package/mongo-dot-notation) arrays are kept, and that means in your case array will be saved as an array, with objects not been changed or flatten within arrays.
That also means that array properties will always be updated entirely, no merging among items within array is done.

**Comment by vcarel** [comment]
Hi, I'm working with @Thebigbignooby and the version 1.0.7 didn't fix the issue...

**Comment by vcarel** [comment]
BUT, since you said that the issue was probably related to null constructors, I tried to deep copy the source object with lodash, and it works !

    dot.flatten(cloneDeep(info))

**Comment by dimadeveatii** [comment]
Issue has been fixed in version 1.0.8.
There is a unit test that covers this scenario.
